src/quemb/kbe/pbe.py

```diff
@@ -12,12 +12,18 @@
 from pyscf.pbc import df, gto
 from pyscf.pbc.df.df_jk import _ewald_exxdiv_for_G0
 
-from quemb.kbe.misc import storePBE
+from quemb.kbe.lo import Mixin_k_Localize
+from quemb.kbe.misc import print_energy, storePBE
 from quemb.kbe.pfrag import Frags
+from quemb.molbe._opt import BEOPT
 from quemb.shared import be_var
+from quemb.shared.external.optqn import (
+    get_be_error_jacobian as _ext_get_be_error_jacobian,
+)
+from quemb.shared.helper import copy_docstring
 
 
-class BE:
+class BE(Mixin_k_Localize):
```

Comment:
On a conceptual level, I am a bit confused why `pbe.BE` class would inherit `Mixin_k_Localize`. I know that you are trying to avoid imports that become class methods, but is this conceptually clearer than maybe creating a localized orbital class that `pbe.BE` object initializes within its constructor and keeps as its attribute?

Reply:
(The following should also answer your [question](https://github.com/troyvvgroup/quemb/pull/30#pullrequestreview-2471580161) about the `Mixin` name )

The full point of a "Mixin" class is to provide method(s) to another class, usually via inheritance, while the "Mixin" class itself cannot be instantiated. 

See a better description [here](https://realpython.com/inheritance-composition-python/#mixing-features-with-mixin-classes), [here](https://stackoverflow.com/questions/533631/what-is-a-mixin-and-why-is-it-useful), or [here](https://medium.com/@yanxingyang/usage-of-mixin-class-in-python-932b940db80#id_token=eyJhbGciOiJSUzI1NiIsImtpZCI6IjJjOGEyMGFmN2ZjOThmOTdmNDRiMTQyYjRkNWQwODg0ZWIwOTM3YzQiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJodHRwczovL2FjY291bnRzLmdvb2dsZS5jb20iLCJhenAiOiIyMTYyOTYwMzU4MzQtazFrNnFlMDYwczJ0cDJhMmphbTRsamRjbXMwMHN0dGcuYXBwcy5nb29nbGV1c2VyY29udGVudC5jb20iLCJhdWQiOiIyMTYyOTYwMzU4MzQtazFrNnFlMDYwczJ0cDJhMmphbTRsamRjbXMwMHN0dGcuYXBwcy5nb29nbGV1c2VyY29udGVudC5jb20iLCJzdWIiOiIxMTU3MzY2NDMzMTk3NDg2ODM4NzIiLCJlbWFpbCI6Im9za2FyLndlc2VyQGdtYWlsLmNvbSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJuYmYiOjE3MzMxNzkzODUsIm5hbWUiOiJPc2thciBXZXNlciIsInBpY3R1cmUiOiJodHRwczovL2xoMy5nb29nbGV1c2VyY29udGVudC5jb20vYS9BQ2c4b2NMV0RzbVhCTV9EalluSW1nVVlWWnNrOXczUFJxQVZ3empJSWVyNXNqRnFHZUg1OVE9czk2LWMiLCJnaXZlbl9uYW1lIjoiT3NrYXIiLCJmYW1pbHlfbmFtZSI6Ildlc2VyIiwiaWF0IjoxNzMzMTc5Njg1LCJleHAiOjE3MzMxODMyODUsImp0aSI6IjBlNjUzYWViY2MzZmNmNDA0ZTE3N2EyZGNhZGE4MjBlZDlkNTJmZTcifQ.fNQbcNL51LhJZeROQgeBe4cUg6FjjU3fVBOOe58I-qN7NioXmSuBcBoceM_F0R31SOiaBgPcLlsP7xuMfoLg0wVDhkutuE5NemaHq37T-98IUmhN__X6XKl3wbdw_XOuZhuCl_BJlXulb150vwQ_i8gia0XnqcLVuvhJbMqSVlkCcufV_wvRQSOXDRkNP-p6Y2xJ1M-q0OcLocLbBvlg6KhyNY7oWrFMvKlk4okM2EdnFOZBRN2_f2E3kiXop1ZObp-Q7RZGKZ9n_qb7-n4u1_pr5mX0tnm95nwutILQHus6-WSTd0pgVPWjQwWyP5WnoAuanB_OkG4-6C5EsUmZFA).


Having a "free" function, that has the argument `self` and actually uses attributes from `self`, is basically the same as having a Mixin class with exactly one method. 
You cannot use this "free" function as a free function on its own.
Modelling this explicitly as a Mixin, makes it a bit more streamlined with general practice. See for example how `pandas` does it [here](https://github.com/pandas-dev/pandas/blob/main/pandas/io/formats/info.py#L841)

> but is this conceptually clearer than maybe creating a localized orbital class that pbe.BE object initializes within its constructor and keeps as its attribute?

I fully agree with your question in general, but it assumes that we can actually meaningfully instantiate a `Localize` class. Not that this is impossible, but at the moment we just have this fake free function and would have to properly refactor. For example we don't want components to show up in both the `BE` and such a `Localize` class.

In the **particular** case of the localize methods  I can actually refactor it away from a Mixin and move it into the `BE` class definition body. It is only imported and used in the `BE` class.

Reply:
Oh, I see. Mixins sound similar to abstract classes in C++. Your choice makes much more sense to me now, and I agree with your reasoning. I think it is early to decide whether it makes sense to refactor out localization methods or not. Assuming we revisit this question at a later point (perhaps create an issue for now?), I approve this PR.